src/drivers/imx/sdma.c

```diff
@@ -730,47 +728,38 @@ static int sdma_prep_desc(struct dma_chan_data *channel,
 		 * For MEM2MEM the source is stored in buf_addr and destination
 		 * is in buf_xaddr.
 		 */
-		bd->buf_xaddr = 0;
-		bd->buf_addr = config->elem_array.elems[i].src;
-		if (!src_may_change)
-			bd->buf_addr =
-				config->elem_array.elems[i].dest;
-		if (src_may_change && dst_may_change) {
-			/* M2M copy */
-			bd->buf_xaddr =
-				config->elem_array.elems[i].dest;
-		}
-
-		/* We are interested in the element width on the hardware side;
-		 * on RAM side we can adjust.
-		 */
-		width = config->src_width;
-		if (!dst_may_change)
+		switch (config->direction) {
+		case DMA_DIR_MEM_TO_DEV:
+			bd->buf_addr = config->elem_array.elems[i].src;
+			width = config->src_width;
+			break;
+		case DMA_DIR_DEV_TO_MEM:
+			bd->buf_addr = config->elem_array.elems[i].dest;
 			width = config->dest_width;
+			break;
+		case DMA_DIR_MEM_TO_MEM:
+			bd->buf_addr = config->elem_array.elems[i].src;
+			bd->buf_xaddr = config->elem_array.elems[i].dest;
+			width = config->dest_width;
+			break;
+		default:
+			return -EINVAL;
+		}
 
 		bd->config =
 			SDMA_BD_COUNT(config->elem_array.elems[i].size) |
 			SDMA_BD_CMD(SDMA_CMD_XFER_SIZE(width)) | SDMA_BD_CONT;
+
 		if (!config->irq_disabled)
 			bd->config |= SDMA_BD_INT;
-		if (dst_may_change) {
-			/* Capture or M2M, enable this descriptor to be
-			 * used by SDMAC
-			 */
-			bd->config |= SDMA_BD_DONE;
-			/* On playback we don't do this and instead wait
-			 * for copy() to let us know the data is ready.
-			 * copy() is called during preload.
-			 */
-		}
-		if (src_may_change && dst_may_change)
-			bd->config |= SDMA_BD_EXTD;
+
+		bd->config |= SDMA_BD_DONE | SDMA_BD_CONT;
 	}
+
 	/* Configure last BD to account for cyclic transfers */
 	if (config->cyclic)
 		bd->config |= SDMA_BD_WRAP;
-	else
-		bd->config &= ~SDMA_BD_CONT;
+	bd->config &= ~SDMA_BD_CONT;
```

Comment:
Should CONT be unset always? This is a behavior change.

Reply:
What happens if CONT is unset for the first BD? Would the second BD still be executed. I think no.

Reply:
Here you unconditionally unset it for the last descriptor and I'm asking why it is so.

Reply:
Becauase last descriptor has BD_WRAP, so when the channel is scheduled again, it will start with the first descriptor. Lets have a talk tomorrow.

Reply:
Last descriptor only has BD_WRAP in the cyclic case?

Reply:
@paulstelian97 yes. Quoting the RM:

```
If the CONTinous bit and Wrap bits are both set in the 3 rd BD, the script will close it and
it will try to open the first BD. An error may occur at this point if the BD#1 has already
been processed and its Done bit is 0. The SDMA script cannot process a BD with a Done
bit to 0. It means the BD is not ready to be processed. To avoid this situation, the
CONTinous bit should not be set for the last BD if Wrap is set, and the Interrupt flag
must set for the last BD. It will warn the owner of the BD that all the BDs have been
processed and it has to re-set to 1 the Done bit of all the BD's if it desires the SDMA to
fill them again
```

Reply:
Well, then I think the proper semantics is to not have a CONT bit set at all, just the DONE and WRAP bits according to what we have in the buffers, and INT on both of them, and CONT would only be used if we will ever do something like that change to EDMA where the descriptor was further split up into many pieces instead of just 2 -- we would have CONT set on all but the last BD that map to an input elem, INT set on the last said BD.

So I guess we should actually completely get rid of CONT for now.

Reply:
Well, that's the problem. I think that without CONT you will never go past the current BD. But I will need to test that.